Sources/xcodeproj/ReferenceGenerator.swift

```diff
@@ -0,0 +1,38 @@
+import Foundation
+
+/// Generates unique deterministic reference. Must be reset if generating for a new project if totally deterministc references are wanted
+public class ReferenceGenerator {
+
+    static let shared = ReferenceGenerator()
+
+    private var references: Set<String> = []
+
+    public init() {
+
+    }
+
+    /// This resets all the cached references to again make them deterministic
+    public func reset() {
+        references = []
+    }
+
+    /// This generates a unique 20 character reference id to be used in ProjectElement.reference
+    ///
+    /// - Parameters:
+    ///   - element: The type of Project Element to generate for
+    ///   - uniqueId: some string value that uniquely defines the object.
+    ///         It doesn't have to be completely unique as duplicates will be handled, but it provides deterministic references
+    /// - Returns: The generated unique reference
+    public func generateReference<T: ProjectElement>(_ element: T.Type, _ uniqueId: String) -> String {
+        var reference = ""
+        var counter = 0
+        let className = String(describing: T.self).replacingOccurrences(of: "PBX", with: "")
```

Comment:
There are some other files that don't start with `PBX`, for example `XCConfigurationList`. Should we also consider them here?